src/main/java/com/cloudbees/jenkins/plugins/kubernetes_credentials_provider/KubernetesCredentialProvider.java

```diff
@@ -69,6 +72,9 @@
     /** Map of Credentials keyed by their credential ID */
     private ConcurrentHashMap<String, IdCredentials> credentials = new ConcurrentHashMap<>();
 
+    /** Map of Credentials to ItemGroup paths they are available in, keyed by credential ID */
+    private Multimap<String, String> credentialScopes = Multimaps.synchronizedListMultimap(ArrayListMultimap.create());
```

Comment:
synchronized maps will serialise all credential lookup. to a single thread, please avoid

Reply:
Since a whole credientialsScopes entry can be created atomically, I'd go for a `ConcurrentHashMap<String, Set<String>>` here (impl with LinkedHashSet if order is of importance).